File: src/main/services/history.ts
/**
 * Per-project chat history persistence so a conversation survives app restarts.
 *
 * The Copilot CLI session id is persisted on the project (see store.ts), which
 * preserves the *agent's* memory; this module persists the *UI transcript* so
 * the two stay in sync. Each project's messages are stored in their own JSON
 * file under `userData/chats/` to avoid rewriting the whole app state (and to
 * keep large transcripts out of the main store file).
 */

import { app } from 'electron'
import { join } from 'path'
import { existsSync, mkdirSync, readFileSync, writeFileSync, rmSync } from 'fs'
import type { ChatMessage } from '../../shared/ipc'

/** Keep transcripts bounded; older messages beyond this are dropped on save. */
const MAX_MESSAGES = 1000

function chatsDir(): string {
  return join(app.getPath('userData'), 'chats')
}

/** Restrict the on-disk filename to a safe slug derived from the project id. */
function historyFile(projectId: string): string {
  const safe = projectId.replace(/[^a-zA-Z0-9_-]/g, '').slice(0, 128)
  return join(chatsDir(), `${safe || 'unknown'}.json`)
}

/** Coerce arbitrary persisted JSON into a clean ChatMessage[]. */
function sanitize(input: unknown): ChatMessage[] {
  if (!Array.isArray(input)) return []
  const out: ChatMessage[] = []
  for (const item of input) {
    const m = item as Partial<ChatMessage>
    if (!m || (m.role !== 'user' && m.role !== 'assistant')) continue
    out.push({
      id: String(m.id ?? ''),
      role: m.role,
      text: typeof m.text === 'string' ? m.text : '',
      tools: Array.isArray(m.tools) ? m.tools : [],
      error: typeof m.error === 'string' ? m.error : undefined,
      attachments: typeof m.attachments === 'number' ? m.attachments : undefined
    })
  }
  return out
}

/** Load a project's persisted conversation (empty array when none/invalid). */
export function loadHistory(projectId: string): ChatMessage[] {
  try {
    return sanitize(JSON.parse(readFileSync(historyFile(projectId), 'utf8')))
  } catch {
    return []
  }
}

/** Persist a project's conversation. An empty array removes the file. */
export function saveHistory(projectId: string, messages: ChatMessage[]): void {
  const clean = sanitize(messages).slice(-MAX_MESSAGES)
  const file = historyFile(projectId)
  if (clean.length === 0) {
    try {
      rmSync(file, { force: true })
    } catch {
      /* best-effort */
    }
    return
  }
  const dir = chatsDir()
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true })
  writeFileSync(file, JSON.stringify(clean), 'utf8')
}

/** Delete a project's persisted conversation (used when a project is removed). */
export function clearHistory(projectId: string): void {
  try {
    rmSync(historyFile(projectId), { force: true })
  } catch {
    /* best-effort */
  }
}

File: src-tauri/src/services/history.rs
//! Per-project chat-history persistence — the Rust counterpart to
//! `src/main/services/history.ts`. Each project's transcript is a JSON file
//! `<projectId>.json` under `<dataDir>/chats/`.

use super::paths;
use crate::types::ChatMessage;

/// Keep transcripts bounded; older messages beyond this are dropped on save.
const MAX_MESSAGES: usize = 1000;

fn safe_slug(input: &str, fallback: &str) -> String {
  let s: String = input
    .chars()
    .filter(|c| c.is_ascii_alphanumeric() || *c == '_' || *c == '-')
    .take(128)
    .collect();
  if s.is_empty() {
    fallback.to_string()
  } else {
    s
  }
}

fn history_file(project_id: &str) -> std::path::PathBuf {
  paths::chats_dir().join(format!("{}.json", safe_slug(project_id, "unknown")))
}

/// Coerce arbitrary persisted JSON into a clean `Vec<ChatMessage>`, dropping
/// legacy "merge" system events from the removed side-threads feature.
fn sanitize(messages: Vec<ChatMessage>) -> Vec<ChatMessage> {
  messages
    .into_iter()
    .filter(|m| (m.role == "user" || m.role == "assistant") && m.kind.as_deref() != Some("merge"))
    .collect()
}

/// Load a project's persisted conversation (empty when none/invalid).
pub fn load_history(project_id: &str) -> Vec<ChatMessage> {
  match std::fs::read_to_string(history_file(project_id)) {
    Ok(raw) => serde_json::from_str::<Vec<ChatMessage>>(&raw)
      .map(sanitize)
      .unwrap_or_default(),
    Err(_) => vec![],
  }
}

/// Persist a project's conversation. An empty list removes the file.
pub fn save_history(project_id: &str, messages: Vec<ChatMessage>) {
  let mut clean = sanitize(messages);
  if clean.len() > MAX_MESSAGES {
    clean = clean.split_off(clean.len() - MAX_MESSAGES);
  }
  let file = history_file(project_id);
  if clean.is_empty() {
    let _ = std::fs::remove_file(&file);
    return;
  }
  if let Err(e) = std::fs::create_dir_all(paths::chats_dir()) {
    log::warn!("failed to create chats dir {}: {e}", paths::chats_dir().display());
  }
  if let Ok(text) = serde_json::to_string(&clean) {
    if let Err(e) = std::fs::write(&file, text) {
      log::warn!("failed to save chat history to {}: {e}", file.display());
    }
  }
}

/// Delete a project's persisted conversation (used on removal).
pub fn clear_history(project_id: &str) {
  let _ = std::fs::remove_file(history_file(project_id));
}
